### ISSUE TITLE
Error: Cannot find module 'jsdom'

### ISSUE THREAD

**Issue by dchubad** [cure53/DOMPurify#475]
Im running node.js on windows 10.
I followed the directions in the readme file for installing DOMPurify server-side but I am presented with this error when ever I connect to my server...

PS C:\xampp\server> node index.js
internal/modules/cjs/loader.js:896
  throw err;
  ^

Error: Cannot find module 'jsdom'
Require stack:
- C:\xampp\server\index.js
←[90m    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:893:15)←[39m
←[90m    at Function.Module._load (internal/modules/cjs/loader.js:743:27)←[39m
←[90m    at Module.require (internal/modules/cjs/loader.js:965:19)←[39m
←[90m    at require (internal/modules/cjs/helpers.js:88:18)←[39m
    at WebSocket.<anonymous> (C:\xampp\server\index.js:13:19)
←[90m    at WebSocket.emit (events.js:314:20)←[39m
    at Receiver.receiverOnMessage (C:\xampp\server\node_modules\←[4mws←[24m\lib\websocket.js:797:20)
←[90m    at Receiver.emit (events.js:314:20)←[39m
    at Receiver.dataMessage (C:\xampp\server\node_modules\←[4mws←[24m\lib\receiver.js:437:14)
    at Receiver.getData (C:\xampp\server\node_modules\←[4mws←[24m\lib\receiver.js:367:17) {
  code: ←[32m'MODULE_NOT_FOUND'←[39m,
  requireStack: [ ←[32m'C:\\xampp\\server\\index.js'←[39m ]
}
PS C:\xampp\server>

**Comment by cure53** [comment]
Can you show how dompurify and jsdom were included?

**Comment by dchubad** [comment]
I just opened cmd, navigated to my usual server dir and entered 'npm install dompurify', there is no mention anywhere how to install jsdom and I asumed installing dompurity would at very least create folders for jadom but there is nothing.
The code below is how I implemented the other code provided by the 'readme' scripts..

```
s.on('connection', function(ws) {
	ws.on('message', function(message){
const createDOMPurify = require('dompurify');
const { JSDOM } = require('jsdom');

const window = new JSDOM('').window;
const DOMPurify = createDOMPurify(window);

		// START only on connection
		message = JSON.parse(message);
		if(message.type == "name"){
		
			// start check for double login
			var dirty = message.data;
			const ConnectingUser = DOMPurify.sanitize(dirty);
```

I am new to node.js and have only 'just' started including addon scripts such as DOMPurify. U'd think they'd already have something like this included in the ws scripts with some configurable options. I do like the way the demo works though, it is perfect for what I want. thanks

**Comment by cure53** [comment]
tbh, I know little about that - our target architecture is the browser :D we just happen to work with jsdom as well.

@tdeekens do you per chance have any idea who to solve this? Likely trivial.

**Comment by dchubad** [comment]
it installs the DOMPurity folder and files in the node_modules folder but I do not see any files related to jsdom.

**Comment by tdeekens** [comment]
From which directory are you running the npm install commands for dompurify and jsdom. Both should be installed from the same root directory if your project I‘d assume. You can also run npm ls to view what has been installed.

**Comment by dchubad** [comment]
c:/xampp/chatserver/

**Comment by dchubad** [comment]
I reinstalled npm and started over again fresh. In the node_modules dir there are several instaled packages which is ws, DOMPurify and a few other packages. This is what my root dir looks like...

C:\xampp\chatserver>dir
 Volume in drive C has no label.
 Volume Serial Number is 34BE-7D50

 Directory of C:\xampp\chatserver

12/10/2020  22:55    <DIR>          .
12/10/2020  22:55    <DIR>          ..
06/10/2020  21:25            29,837 chatroom.js
12/10/2020  09:31            30,069 index.js
12/10/2020  22:55    <DIR>          node_modules
12/10/2020  22:55               594 package-lock.json
12/10/2020  22:55               288 package.json
               7 File(s)         62,788 bytes
               5 Dir(s)  244,010,749,952 bytes free

C:\xampp\chatserver>node chatroom.js

In the directory node_modules there are these directories...

 Directory of C:\xampp\chatserver\node_modules

13/10/2020  00:02    <DIR>          .
13/10/2020  00:02    <DIR>          ..
13/10/2020  00:02    <DIR>          .bin
13/10/2020  00:02    <DIR>          ansi-regex
13/10/2020  00:02    <DIR>          async
13/10/2020  00:02    <DIR>          camelcase
13/10/2020  00:02    <DIR>          cliui
13/10/2020  00:02    <DIR>          code-point-at
13/10/2020  00:02    <DIR>          decamelize
12/10/2020  22:55    <DIR>          dompurify
13/10/2020  00:02    <DIR>          ini
13/10/2020  00:02    <DIR>          invert-kv
13/10/2020  00:02    <DIR>          is-fullwidth-code-point
13/10/2020  00:02    <DIR>          lcid
13/10/2020  00:02    <DIR>          nconf
13/10/2020  00:02    <DIR>          number-is-nan
13/10/2020  00:02    <DIR>          os-locale
13/10/2020  00:02    <DIR>          secure-keys
13/10/2020  00:02    <DIR>          string-width
13/10/2020  00:02    <DIR>          strip-ansi
13/10/2020  00:02    <DIR>          window-size
13/10/2020  00:02    <DIR>          wrap-ansi
12/10/2020  22:48    <DIR>          ws
13/10/2020  00:02    <DIR>          y18n
13/10/2020  00:02    <DIR>          yargs
               0 File(s)              0 bytes
              25 Dir(s)  244,010,516,480 bytes free

not sure if this helps.

**Comment by tdeekens** [comment]
I fail to understand why `jsdom` is not there if `npm install jsdom` was run from the `chatserver` directory. I also don't think it's an issue point to DOMPurify itself.

**Comment by dchubad** [comment]
I finally got it working, for some reason when I closed powershell it would not open again. I also realised a few days before I updated my system from microsoft and failed to restart as requested but continued with trying to get this working. After a restart I managed to open powershell, entered npm install jsdom and the final files were installed successfuly. I will not get time to test it fully until later but for now it is working, no errors when text is entered into chat. Will test it out fully when I get home later but thanks for your help. Sorry if I wasted anyones time, it was not my intention.
